src/plugins/chart_expressions/expression_xy/public/helpers/data_layers.tsx

```diff
@@ -0,0 +1,302 @@
+  ScaleType,
+  SeriesName,
+  StackMode,
+  XYChartSeriesIdentifier,
+} from '@elastic/charts';
+import { i18n } from '@kbn/i18n';
+import {
+  FieldFormat,
+  FieldFormatParams,
+  SerializedFieldFormat,
+} from '@kbn/field-formats-plugin/common';
+import { Datatable, DatatableRow } from '@kbn/expressions-plugin';
+import { PaletteRegistry, SeriesLayer } from '@kbn/coloring';
+import { CommonXYDataLayerConfig, XScaleType } from '../../common';
+import { FormatFactory } from '../types';
+import { getSeriesColor } from './state';
+import { ColorAssignments } from './color_assignment';
+import { GroupsConfiguration } from './axes_configuration';
+
+type SeriesSpec = LineSeriesProps & BarSeriesProps & AreaSeriesProps;
+
+type GetSeriesPropsFn = (config: {
+  layer: CommonXYDataLayerConfig;
+  accessor: string;
+  chartHasMoreThanOneBarSeries?: boolean;
+  formatFactory: FormatFactory;
+  colorAssignments: ColorAssignments;
+  columnToLabelMap: Record<string, string>;
+  paletteService: PaletteRegistry;
+  alreadyFormattedColumns: Record<string, boolean>;
+  syncColors?: boolean;
+  yAxis?: GroupsConfiguration[number];
+  timeZone?: string;
+  emphasizeFitting?: boolean;
+  fillOpacity?: number;
+}) => SeriesSpec;
+
+type GetSeriesNameFn = (
+  data: XYChartSeriesIdentifier,
+  config: {
+    layer: CommonXYDataLayerConfig;
+    splitHint: SerializedFieldFormat<FieldFormatParams> | undefined;
+    splitFormatter: FieldFormat;
+    alreadyFormattedColumns: Record<string, boolean>;
+    columnToLabelMap: Record<string, string>;
+  }
+) => SeriesName;
+
+type GetColorFn = (
+  seriesIdentifier: XYChartSeriesIdentifier,
+  config: {
+    layer: CommonXYDataLayerConfig;
+    accessor: string;
+    colorAssignments: ColorAssignments;
+    columnToLabelMap: Record<string, string>;
+    paletteService: PaletteRegistry;
+    syncColors?: boolean;
+  }
+) => string | null;
+
+const isPrimitive = (value: unknown): boolean => value != null && typeof value !== 'object';
+
+export const getFormattedTable = (
+  table: Datatable,
+  formatFactory: FormatFactory,
+  xAccessor: string | undefined,
+  xScaleType: XScaleType
+): Datatable => ({
+  ...table,
+  rows: table.rows.map((row: DatatableRow) => {
+    const newRow = { ...row };
+    for (const column of table.columns) {
+      const record = newRow[column.id];
+      if (
+        record != null &&
+        // pre-format values for ordinal x axes because there can only be a single x axis formatter on chart level
+        (!isPrimitive(record) || (column.id === xAccessor && xScaleType === 'ordinal'))
+      ) {
+        newRow[column.id] = formatFactory(column.meta.params)!.convert(record);
+      }
+    }
+    return newRow;
+  }),
+});
+
+export const getIsAlreadyFormattedLayerInfo = (
+  { table, xAccessor, xScaleType }: CommonXYDataLayerConfig,
+  formatFactory: FormatFactory
+): Record<string, boolean> => {
+  const formattedTable = getFormattedTable(table, formatFactory, xAccessor, xScaleType);
```

Comment:
This new structure is formatting the table twice - once for calculating the `isAlreadyFormatted` object, then again a second time to calculate the series props for the data layers. We should only call `getFormattedTable` once, then pass it to both places. Formatting is not cheap is it is called for the every cell of the table which can be quite large).

Reply:
Completely changed this approach. Pushed changes.

Reply:
Done.